ansible_base/authentication/utils/authentication.py

```diff
@@ -20,6 +21,57 @@ def setting(self, *args, **kwargs):
         return ["username", "email"]
 
 
+def migrate_from_existing_authenticator(
+    uid: str, alt_uid: Optional[str], authenticator: Authenticator, preferred_username: Optional[str] = None
+) -> Optional[str]:
+    """
+    uid: the users uid.
+    alt_uid: an optional alternative uid to use for looking up other accounts.
+    authenticator: the authenticator that the user is currently authenticating with.
+    preferred_username: the username that the authenticator wants to use for the user.
+
+    Returns the username of the django account to user for the authenticated user, or None if no match was found.
+    """
+
+    # SAML puts prepends all the UIDs with IdP. Adding this to the search criteria will
+    # allow us to find SAML accounts that match the UID.
+    uid_filter = [uid, "IdP:" + uid]
+    if alt_uid:
+        uid_filter.append(alt_uid)
+
+    migrate_users = list(
+        AuthenticatorUser.objects.filter(
+            uid__in=uid_filter,
+            provider__auto_migrate_users_to=authenticator.pk,
+        ).order_by("provider__order")
+    )
+
+    if len(migrate_users) == 0:
+        return
+
+    try:
+        main_user = AuthenticatorUser.objects.get(uid=uid, provider=authenticator).user
+    except AuthenticatorUser.DoesNotExist:
+        main_user = migrate_users[0].user
+
+    for migrate_user in migrate_users:
+        provider = migrate_user.provider
+        from_authenticator = get_authenticator_class(provider.type)(database_instance=provider)
+        old_user = from_authenticator.move_authenticator_user_to(main_user, migrate_user)
+        if old_user and not old_user.authenticator_users.exists():
```

Comment:
If the `old_user` is still associated with other, unrelated, authenticators, it will have already had its associations applied to the new user. But reading more closely, it seems that you (reasonably) did not remove those associations. That is probably the most correct behavior. But I think you should issue a warning log in an `else:` here because that's a fairly degenerate state.

I also see that you're not copying `role_assignments` from the prior file... so how are those ever going to get migrated? Because this logic seems to drop them on the floor.